packages/flutter/lib/src/material/list_tile.dart

```diff
@@ -385,91 +404,553 @@ class ListTile extends StatelessWidget {
 
+    Widget leadingIcon;
     if (leading != null) {
-      children.add(IconTheme.merge(
+      leadingIcon = IconTheme.merge(
         data: iconThemeData,
-        child: new Container(
-          margin: const EdgeInsetsDirectional.only(end: 16.0),
-          width: 40.0,
-          alignment: AlignmentDirectional.centerStart,
-          child: leading,
-        ),
-      ));
+        child: leading,
+      );
     }
 
-    final Widget primaryLine = new AnimatedDefaultTextStyle(
+    final Widget titleText = new AnimatedDefaultTextStyle(
       style: _titleTextStyle(theme, tileTheme),
       duration: kThemeChangeDuration,
-      child: title ?? new Container()
+      child: title ?? const SizedBox()
     );
-    Widget center = primaryLine;
-    if (subtitle != null && (isTwoLine || isThreeLine)) {
-      center = new Column(
-        mainAxisSize: MainAxisSize.min,
-        crossAxisAlignment: CrossAxisAlignment.start,
-        children: <Widget>[
-          primaryLine,
-          new AnimatedDefaultTextStyle(
-            style: _subtitleTextStyle(theme, tileTheme),
-            duration: kThemeChangeDuration,
-            child: subtitle,
-          ),
-        ],
+
+    Widget subtitleText;
+    if (subtitle != null) {
+      subtitleText = new AnimatedDefaultTextStyle(
+        style: _subtitleTextStyle(theme, tileTheme),
+        duration: kThemeChangeDuration,
+        child: subtitle,
       );
     }
-    children.add(new Expanded(
-      child: center,
-    ));
 
+    Widget trailingIcon;
     if (trailing != null) {
-      children.add(IconTheme.merge(
+      trailingIcon = IconTheme.merge(
         data: iconThemeData,
-        child: new Container(
-          margin: const EdgeInsetsDirectional.only(start: 16.0),
-          alignment: AlignmentDirectional.centerEnd,
-          child: trailing,
-        ),
-      ));
+        child: trailing,
+      );
     }
 
+    final TextDirection textDirection = Directionality.of(context);
+    final EdgeInsets resolvedContentPadding = contentPadding?.resolve(textDirection)
+      ?? tileTheme?.contentPadding?.resolve(textDirection)
+      ?? const EdgeInsets.symmetric(horizontal: 16.0);
+
     return new InkWell(
       onTap: enabled ? onTap : null,
       onLongPress: enabled ? onLongPress : null,
       child: new Semantics(
         selected: selected,
         enabled: enabled,
-        child: new ConstrainedBox(
-          constraints: new BoxConstraints(minHeight: tileHeight),
-          child: new Padding(
-            padding: const EdgeInsets.symmetric(horizontal: 16.0),
-            child: new UnconstrainedBox(
-              constrainedAxis: Axis.horizontal,
-              child: new SafeArea(
-                top: false,
-                bottom: false,
-                child: new Row(children: children),
-              ),
+        child: new Padding(
+          padding: resolvedContentPadding,
+          child: new SafeArea(
```

Comment:
Wouldn't it be better to have the safe area outside of the padding?  This way it might look a little close to a side notch if it's at the edge.  And why only have it take effect horizontally?

Reply:
Good idea, will do that.